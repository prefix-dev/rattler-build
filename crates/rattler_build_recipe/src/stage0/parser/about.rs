//! Parser for the About section

use marked_yaml::Node as MarkedNode;

use crate::{
    error::{ParseError, ParseResult},
    span::SpannedString,
    stage0::{
        about::About,
        parser::{helpers::get_span, list::parse_conditional_list, value::parse_value},
        types::{ConditionalList, Item, JinjaTemplate, Value},
    },
};

/// Parse license_file field which can be either a single string or a list of strings
fn parse_license_file(yaml: &MarkedNode) -> ParseResult<ConditionalList<String>> {
    // Try parsing as a sequence first
    if yaml.as_sequence().is_some() {
        return parse_conditional_list(yaml);
    }

    // Try parsing as a single scalar string
    if let Some(scalar) = yaml.as_scalar() {
        let spanned = SpannedString::from(scalar);
        let s = spanned.as_str();

        // Check if it's a template
        if s.contains("${{") && s.contains("}}") {
            let template = JinjaTemplate::new(s.to_string())
                .map_err(|e| ParseError::jinja_error(e, spanned.span()))?;
            let items = vec![Item::Value(Value::Template(template))];
            return Ok(ConditionalList::new(items));
        }

        // Plain string
        let items = vec![Item::Value(Value::Concrete(s.to_string()))];
        return Ok(ConditionalList::new(items));
    }

    // Get proper span for better error message
    let span = get_span(yaml);
    Err(
        ParseError::expected_type("string or list of strings", "other", span)
            .with_message("license_file must be a string or a list of strings"),
    )
}

/// Parse an About section from YAML
///
/// All fields in the About section are optional and can contain templates.
///
/// Example YAML:
/// ```yaml
/// about:
///   homepage: https://example.com
///   license: MIT
///   license_file: LICENSE
///   summary: A short description
///   description: A longer description
///   documentation: https://docs.example.com
///   repository: https://github.com/example/repo
/// ```
pub fn parse_about(yaml: &MarkedNode) -> ParseResult<About> {
    let mapping = yaml.as_mapping().ok_or_else(|| {
        ParseError::expected_type("mapping", "non-mapping", get_span(yaml))
            .with_message("About section must be a mapping")
    })?;

    let mut about = About::default();

    // Parse each optional field
    if let Some(homepage) = mapping.get("homepage") {
        about.homepage = Some(parse_value(homepage)?);
    }

    if let Some(license) = mapping.get("license") {
        about.license = Some(parse_value(license)?);
    }

    if let Some(license_file) = mapping.get("license_file") {
        about.license_file = parse_license_file(license_file)?;
    }

    if let Some(license_family) = mapping.get("license_family") {
        about.license_family = Some(parse_value(license_family)?);
    }

    if let Some(summary) = mapping.get("summary") {
        about.summary = Some(parse_value(summary)?);
    }

    if let Some(description) = mapping.get("description") {
        about.description = Some(parse_value(description)?);
    }

    if let Some(documentation) = mapping.get("documentation") {
        about.documentation = Some(parse_value(documentation)?);
    }

    if let Some(repository) = mapping.get("repository") {
        about.repository = Some(parse_value(repository)?);
    }

    // Check for unknown fields to provide helpful error messages
    for (key, _) in mapping.iter() {
        let key_str = key.as_str();
        if !matches!(
            key_str,
            "homepage"
                | "license"
                | "license_file"
                | "license_family"
                | "summary"
                | "description"
                | "documentation"
                | "repository"
        ) {
            return Err(ParseError::invalid_value(
                "about",
                &format!("unknown field '{}'", key_str),
                (*key.span()).into(),
            )
            .with_suggestion(
                "valid fields are: homepage, license, license_file, license_family, summary, description, documentation, repository"
            ));
        }
    }

    Ok(about)
}

#[cfg(test)]
mod tests {
    use super::*;

    fn parse_yaml_about(yaml_str: &str) -> MarkedNode {
        let wrapped = format!("about:\n{}", yaml_str);
        let root = marked_yaml::parse_yaml(0, &wrapped).expect("Failed to parse test YAML");
        let mapping = root.as_mapping().expect("Expected mapping");
        mapping.get("about").expect("Field not found").clone()
    }

    #[test]
    fn test_parse_empty_about() {
        let yaml = parse_yaml_about("  {}");
        let about = parse_about(&yaml).unwrap();
        assert!(about.homepage.is_none());
        assert!(about.license.is_none());
        assert!(about.summary.is_none());
    }

    #[test]
    fn test_parse_about_with_concrete_values() {
        let yaml_str = r#"
  homepage: https://example.com
  license: MIT
  summary: A test package"#;
        let yaml = parse_yaml_about(yaml_str);
        let about = parse_about(&yaml).unwrap();

        assert!(about.homepage.is_some());
        assert!(about.license.is_some());
        assert!(about.summary.is_some());

        // Verify concrete values
        match about.homepage.as_ref().unwrap() {
            crate::stage0::types::Value::Concrete(url) => {
                assert_eq!(url.as_str(), "https://example.com/");
            }
            _ => panic!("Expected concrete value"),
        }

        match about.license.as_ref().unwrap() {
            crate::stage0::types::Value::Concrete(license) => {
                assert_eq!(license.0.as_ref(), "MIT");
            }
            _ => panic!("Expected concrete value"),
        }
    }

    #[test]
    fn test_parse_about_with_templates() {
        let yaml_str = r#"
  homepage: '${{ homepage }}'
  license: '${{ license }}'
  summary: '${{ name }} - ${{ summary }}'"#;
        let yaml = parse_yaml_about(yaml_str);
        let about = parse_about(&yaml).unwrap();

        // Verify templates
        match about.homepage.as_ref().unwrap() {
            crate::stage0::types::Value::Template(t) => {
                assert_eq!(t.used_variables(), &["homepage"]);
            }
            _ => panic!("Expected template value"),
        }

        match about.summary.as_ref().unwrap() {
            crate::stage0::types::Value::Template(t) => {
                let mut vars = t.used_variables().to_vec();
                vars.sort();
                assert_eq!(vars, vec!["name", "summary"]);
            }
            _ => panic!("Expected template value"),
        }
    }

    #[test]
    fn test_parse_about_all_fields() {
        let yaml_str = r#"
  homepage: https://example.com
  license: Apache-2.0
  license_file: LICENSE
  summary: A test package
  description: A longer description
  documentation: https://docs.example.com
  repository: https://github.com/example/repo"#;
        let yaml = parse_yaml_about(yaml_str);
        let about = parse_about(&yaml).unwrap();

        assert!(about.homepage.is_some());
        assert!(about.license.is_some());
        assert!(!about.license_file.is_empty());
        assert!(about.summary.is_some());
        assert!(about.description.is_some());
        assert!(about.documentation.is_some());
        assert!(about.repository.is_some());
    }

    #[test]
    fn test_parse_about_unknown_field() {
        let yaml_str = r#"
  homepage: https://example.com
  unknown_field: some value"#;
        let yaml = parse_yaml_about(yaml_str);
        let result = parse_about(&yaml);

        assert!(result.is_err());
        let err = result.unwrap_err();
        assert!(err.message.as_ref().unwrap().contains("unknown field"));
    }

    #[test]
    fn test_parse_about_not_mapping() {
        let wrapped = "about: not a mapping";
        let root = marked_yaml::parse_yaml(0, wrapped).expect("Failed to parse test YAML");
        let mapping = root.as_mapping().expect("Expected mapping");
        let yaml = mapping.get("about").expect("Field not found");

        let result = parse_about(yaml);
        assert!(result.is_err());
        let err = result.unwrap_err();
        assert!(err.message.as_ref().unwrap().contains("must be a mapping"));
    }

    #[test]
    fn test_parse_about_partial_fields() {
        let yaml_str = r#"
  license: MIT
  summary: Just a summary"#;
        let yaml = parse_yaml_about(yaml_str);
        let about = parse_about(&yaml).unwrap();

        assert!(about.homepage.is_none());
        assert!(about.license.is_some());
        assert!(about.summary.is_some());
        assert!(about.repository.is_none());
    }
}
