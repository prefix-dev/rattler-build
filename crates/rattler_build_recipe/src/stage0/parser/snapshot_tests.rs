//! Snapshot tests for recipe parsing using insta
//!
//! These tests parse real recipe files from test-data/ and snapshot the results

use crate::stage0::parser::parse_recipe_from_source;

const TEST_DATA_DIR: &str = concat!(env!("CARGO_MANIFEST_DIR"), "/test-data");

fn load_test_recipe(filename: &str) -> String {
    let path = format!("{}/{}", TEST_DATA_DIR, filename);
    fs_err::read_to_string(&path)
        .unwrap_or_else(|e| panic!("Failed to read test file {}: {}", path, e))
}

#[test]
fn test_minimal_recipe_snapshot() {
    let source = load_test_recipe("minimal.yaml");
    let recipe = parse_recipe_from_source(&source).expect("Failed to parse minimal recipe");
    insta::assert_debug_snapshot!(recipe);
}

#[test]
fn test_full_recipe_snapshot() {
    let source = load_test_recipe("full.yaml");
    let recipe = parse_recipe_from_source(&source).expect("Failed to parse full recipe");
    insta::assert_debug_snapshot!(recipe);
}

#[test]
fn test_templates_recipe_snapshot() {
    let source = load_test_recipe("templates.yaml");
    let recipe = parse_recipe_from_source(&source).expect("Failed to parse templates recipe");
    insta::assert_debug_snapshot!(recipe);
}

#[test]
fn test_conditionals_recipe_snapshot() {
    let source = load_test_recipe("conditionals.yaml");
    let recipe = parse_recipe_from_source(&source).expect("Failed to parse conditionals recipe");
    insta::assert_debug_snapshot!(recipe);
}

#[test]
fn test_complex_recipe_snapshot() {
    let source = load_test_recipe("complex.yaml");
    let recipe = parse_recipe_from_source(&source).expect("Failed to parse complex recipe");
    insta::assert_debug_snapshot!(recipe);
}

#[test]
fn test_run_exports_recipe_snapshot() {
    let source = load_test_recipe("run_exports.yaml");
    let recipe = parse_recipe_from_source(&source).expect("Failed to parse run_exports recipe");
    insta::assert_debug_snapshot!(recipe);
}

#[test]
fn test_templates_extract_variables() {
    let source = load_test_recipe("templates.yaml");
    let recipe = parse_recipe_from_source(&source).expect("Failed to parse templates recipe");
    let vars = recipe.used_variables();
    insta::assert_debug_snapshot!(vars);
}

#[test]
fn test_conditionals_extract_variables() {
    let source = load_test_recipe("conditionals.yaml");
    let recipe = parse_recipe_from_source(&source).expect("Failed to parse conditionals recipe");
    let vars = recipe.used_variables();
    insta::assert_debug_snapshot!(vars);
}

#[test]
fn test_complex_extract_variables() {
    let source = load_test_recipe("complex.yaml");
    let recipe = parse_recipe_from_source(&source).expect("Failed to parse complex recipe");
    let vars = recipe.used_variables();
    insta::assert_debug_snapshot!(vars);
}
