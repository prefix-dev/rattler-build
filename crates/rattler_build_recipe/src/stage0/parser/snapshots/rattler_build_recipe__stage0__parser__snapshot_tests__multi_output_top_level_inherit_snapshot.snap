---
source: crates/rattler_build_recipe/src/stage0/parser/snapshot_tests.rs
expression: recipe
---
MultiOutput(
    MultiOutputRecipe {
        schema_version: Some(
            1,
        ),
        context: {},
        recipe: RecipeMetadata {
            name: None,
            version: Some(
                Concrete {
                    value: VersionWithSource {
                        version: Version {
                            version: [[0], [5], [0], [0]],
                            local: [],
                        },
                        source: Some(
                            "5.0.0",
                        ),
                    },
                    span: Span {
                        start: 96,
                        end: 96,
                        start_line: 5,
                        start_column: 12,
                        end_line: 5,
                        end_column: 12,
                    },
                },
            ),
        },
        source: [
            Url(
                UrlSource {
                    url: [
                        Concrete {
                            value: "https://example.com/top-level-source.tar.gz",
                            span: Span {
                                start: 120,
                                end: 120,
                                start_line: 8,
                                start_column: 10,
                                end_line: 8,
                                end_column: 10,
                            },
                        },
                    ],
                    sha256: Some(
                        Concrete {
                            value: [
                                227,
                                176,
                                196,
                                66,
                                152,
                                252,
                                28,
                                20,
                                154,
                                251,
                                244,
                                200,
                                153,
                                111,
                                185,
                                36,
                                39,
                                174,
                                65,
                                228,
                                100,
                                155,
                                147,
                                76,
                                164,
                                149,
                                153,
                                27,
                                120,
                                82,
                                184,
                                85,
                            ],
                            span: Span {
                                start: 176,
                                end: 176,
                                start_line: 9,
                                start_column: 13,
                                end_line: 9,
                                end_column: 13,
                            },
                        },
                    ),
                    md5: None,
                    file_name: None,
                    patches: ConditionalList(
                        [],
                    ),
                    target_directory: None,
                },
            ),
        ],
        build: Build {
            number: Concrete {
                value: 0,
                span: Span {
                    start: 259,
                    end: 259,
                    start_line: 12,
                    start_column: 11,
                    end_line: 12,
                    end_column: 11,
                },
            },
            string: None,
            script: ConditionalList(
                [
                    Value(Concrete { value: Command("echo \"Top-level build script\""), span: Span { start: 277, end: 277, start_line: 14, start_column: 7, end_line: 14, end_column: 7 } }),
                ],
            ),
            noarch: None,
            python: PythonBuild {
                entry_points: ConditionalList(
                    [],
                ),
                skip_pyc_compilation: ConditionalList(
                    [],
                ),
                use_python_app_entrypoint: false,
                version_independent: false,
                site_packages_path: None,
            },
            skip: ConditionalList(
                [],
            ),
            always_copy_files: ConditionalList(
                [],
            ),
            always_include_files: ConditionalList(
                [],
            ),
            merge_build_and_host_envs: false,
            files: List(
                ConditionalList(
                    [],
                ),
            ),
            dynamic_linking: DynamicLinking {
                rpaths: ConditionalList(
                    [],
                ),
                binary_relocation: Boolean(
                    Concrete {
                        value: true,
                        span: Span {
                            start: 0,
                            end: 0,
                            start_line: 0,
                            start_column: 0,
                            end_line: 0,
                            end_column: 0,
                        },
                    },
                ),
                missing_dso_allowlist: ConditionalList(
                    [],
                ),
                rpath_allowlist: ConditionalList(
                    [],
                ),
                overdepending_behavior: None,
                overlinking_behavior: None,
            },
            variant: VariantKeyUsage {
                use_keys: ConditionalList(
                    [],
                ),
                ignore_keys: ConditionalList(
                    [],
                ),
                down_prioritize_variant: None,
            },
            prefix_detection: PrefixDetection {
                force_file_type: ForceFileType {
                    text: ConditionalList(
                        [],
                    ),
                    binary: ConditionalList(
                        [],
                    ),
                },
                ignore: Boolean(
                    Concrete {
                        value: false,
                        span: Span {
                            start: 0,
                            end: 0,
                            start_line: 0,
                            start_column: 0,
                            end_line: 0,
                            end_column: 0,
                        },
                    },
                ),
                ignore_binary_files: false,
            },
            post_process: [],
        },
        about: About {
            homepage: Some(
                Concrete {
                    value: Url {
                        scheme: "https",
                        cannot_be_a_base: false,
                        username: "",
                        password: None,
                        host: Some(
                            Domain(
                                "top-level-example.com",
                            ),
                        ),
                        port: None,
                        path: "/",
                        query: None,
                        fragment: None,
                    },
                    span: Span {
                        start: 355,
                        end: 355,
                        start_line: 18,
                        start_column: 13,
                        end_line: 18,
                        end_column: 13,
                    },
                },
            ),
            license: Some(
                Concrete {
                    value: License(
                        GPL-3.0-or-later,
                    ),
                    span: Span {
                        start: 326,
                        end: 326,
                        start_line: 17,
                        start_column: 12,
                        end_line: 17,
                        end_column: 12,
                    },
                },
            ),
            license_file: ConditionalList(
                [],
            ),
            license_family: None,
            summary: None,
            description: None,
            documentation: None,
            repository: None,
        },
        extra: Extra {
            recipe_maintainers: ConditionalList(
                [],
            ),
        },
        outputs: [
            Package(
                PackageOutput {
                    package: PackageMetadata {
                        name: Concrete {
                            value: PackageName(
                                PackageName {
                                    normalized: None,
                                    source: "main-package",
                                },
                            ),
                            span: Span {
                                start: 477,
                                end: 477,
                                start_line: 23,
                                start_column: 13,
                                end_line: 23,
                                end_column: 13,
                            },
                        },
                        version: Some(
                            Concrete {
                                value: VersionWithSource {
                                    version: Version {
                                        version: [[0], [5], [0], [0]],
                                        local: [],
                                    },
                                    source: Some(
                                        "5.0.0",
                                    ),
                                },
                                span: Span {
                                    start: 505,
                                    end: 505,
                                    start_line: 24,
                                    start_column: 16,
                                    end_line: 24,
                                    end_column: 16,
                                },
                            },
                        ),
                    },
                    inherit: TopLevel,
                    source: [],
                    requirements: Requirements {
                        build: ConditionalList(
                            [],
                        ),
                        host: ConditionalList(
                            [],
                        ),
                        run: ConditionalList(
                            [
                                Value(Concrete { value: SerializableMatchSpec(MatchSpec { name: Some(PackageName { normalized: None, source: "python" }), version: None, build: None, build_number: None, file_name: None, extras: None, channel: None, subdir: None, namespace: None, md5: None, sha256: None, url: None, license: None }), span: Span { start: 568, end: 568, start_line: 28, start_column: 11, end_line: 28, end_column: 11 } }),
                            ],
                        ),
                        run_constraints: ConditionalList(
                            [],
                        ),
                        run_exports: RunExports {
                            noarch: ConditionalList(
                                [],
                            ),
                            strong: ConditionalList(
                                [],
                            ),
                            strong_constraints: ConditionalList(
                                [],
                            ),
                            weak: ConditionalList(
                                [],
                            ),
                            weak_constraints: ConditionalList(
                                [],
                            ),
                        },
                        ignore_run_exports: IgnoreRunExports {
                            by_name: ConditionalList(
                                [],
                            ),
                            from_package: ConditionalList(
                                [],
                            ),
                        },
                    },
                    build: Build {
                        number: Concrete {
                            value: 0,
                            span: Span {
                                start: 0,
                                end: 0,
                                start_line: 0,
                                start_column: 0,
                                end_line: 0,
                                end_column: 0,
                            },
                        },
                        string: None,
                        script: ConditionalList(
                            [],
                        ),
                        noarch: None,
                        python: PythonBuild {
                            entry_points: ConditionalList(
                                [],
                            ),
                            skip_pyc_compilation: ConditionalList(
                                [],
                            ),
                            use_python_app_entrypoint: false,
                            version_independent: false,
                            site_packages_path: None,
                        },
                        skip: ConditionalList(
                            [],
                        ),
                        always_copy_files: ConditionalList(
                            [],
                        ),
                        always_include_files: ConditionalList(
                            [],
                        ),
                        merge_build_and_host_envs: false,
                        files: List(
                            ConditionalList(
                                [],
                            ),
                        ),
                        dynamic_linking: DynamicLinking {
                            rpaths: ConditionalList(
                                [],
                            ),
                            binary_relocation: Boolean(
                                Concrete {
                                    value: true,
                                    span: Span {
                                        start: 0,
                                        end: 0,
                                        start_line: 0,
                                        start_column: 0,
                                        end_line: 0,
                                        end_column: 0,
                                    },
                                },
                            ),
                            missing_dso_allowlist: ConditionalList(
                                [],
                            ),
                            rpath_allowlist: ConditionalList(
                                [],
                            ),
                            overdepending_behavior: None,
                            overlinking_behavior: None,
                        },
                        variant: VariantKeyUsage {
                            use_keys: ConditionalList(
                                [],
                            ),
                            ignore_keys: ConditionalList(
                                [],
                            ),
                            down_prioritize_variant: None,
                        },
                        prefix_detection: PrefixDetection {
                            force_file_type: ForceFileType {
                                text: ConditionalList(
                                    [],
                                ),
                                binary: ConditionalList(
                                    [],
                                ),
                            },
                            ignore: Boolean(
                                Concrete {
                                    value: false,
                                    span: Span {
                                        start: 0,
                                        end: 0,
                                        start_line: 0,
                                        start_column: 0,
                                        end_line: 0,
                                        end_column: 0,
                                    },
                                },
                            ),
                            ignore_binary_files: false,
                        },
                        post_process: [],
                    },
                    about: About {
                        homepage: None,
                        license: None,
                        license_file: ConditionalList(
                            [],
                        ),
                        license_family: None,
                        summary: Some(
                            Concrete {
                                value: "Main package inheriting from top-level",
                                span: Span {
                                    start: 601,
                                    end: 601,
                                    start_line: 30,
                                    start_column: 16,
                                    end_line: 30,
                                    end_column: 16,
                                },
                            },
                        ),
                        description: None,
                        documentation: None,
                        repository: None,
                    },
                    tests: [
                        Python {
                            python: PythonTest {
                                imports: ConditionalList(
                                    [
                                        Value(Concrete { value: "main_package", span: Span { start: 700, end: 700, start_line: 34, start_column: 15, end_line: 34, end_column: 15 } }),
                                    ],
                                ),
                                pip_check: None,
                                python_version: None,
                            },
                        },
                    ],
                },
            ),
            Package(
                PackageOutput {
                    package: PackageMetadata {
                        name: Concrete {
                            value: PackageName(
                                PackageName {
                                    normalized: None,
                                    source: "sub-package",
                                },
                            ),
                            span: Span {
                                start: 819,
                                end: 819,
                                start_line: 38,
                                start_column: 13,
                                end_line: 38,
                                end_column: 13,
                            },
                        },
                        version: Some(
                            Concrete {
                                value: VersionWithSource {
                                    version: Version {
                                        version: [[0], [5], [0], [0]],
                                        local: [],
                                    },
                                    source: Some(
                                        "5.0.0",
                                    ),
                                },
                                span: Span {
                                    start: 846,
                                    end: 846,
                                    start_line: 39,
                                    start_column: 16,
                                    end_line: 39,
                                    end_column: 16,
                                },
                            },
                        ),
                    },
                    inherit: TopLevel,
                    source: [],
                    requirements: Requirements {
                        build: ConditionalList(
                            [],
                        ),
                        host: ConditionalList(
                            [],
                        ),
                        run: ConditionalList(
                            [
                                Value(Concrete { value: SerializableMatchSpec(MatchSpec { name: Some(PackageName { normalized: None, source: "python" }), version: None, build: None, build_number: None, file_name: None, extras: None, channel: None, subdir: None, namespace: None, md5: None, sha256: None, url: None, license: None }), span: Span { start: 891, end: 891, start_line: 42, start_column: 11, end_line: 42, end_column: 11 } }),
                                Value(Concrete { value: SerializableMatchSpec(MatchSpec { name: Some(PackageName { normalized: None, source: "main-package" }), version: Some(Exact(Equals, Version { version: [[0], [5], [0], [0]], local: [] })), build: None, build_number: None, file_name: None, extras: None, channel: None, subdir: None, namespace: None, md5: None, sha256: None, url: None, license: None }), span: Span { start: 908, end: 908, start_line: 43, start_column: 11, end_line: 43, end_column: 11 } }),
                            ],
                        ),
                        run_constraints: ConditionalList(
                            [],
                        ),
                        run_exports: RunExports {
                            noarch: ConditionalList(
                                [],
                            ),
                            strong: ConditionalList(
                                [],
                            ),
                            strong_constraints: ConditionalList(
                                [],
                            ),
                            weak: ConditionalList(
                                [],
                            ),
                            weak_constraints: ConditionalList(
                                [],
                            ),
                        },
                        ignore_run_exports: IgnoreRunExports {
                            by_name: ConditionalList(
                                [],
                            ),
                            from_package: ConditionalList(
                                [],
                            ),
                        },
                    },
                    build: Build {
                        number: Concrete {
                            value: 0,
                            span: Span {
                                start: 0,
                                end: 0,
                                start_line: 0,
                                start_column: 0,
                                end_line: 0,
                                end_column: 0,
                            },
                        },
                        string: None,
                        script: ConditionalList(
                            [],
                        ),
                        noarch: None,
                        python: PythonBuild {
                            entry_points: ConditionalList(
                                [],
                            ),
                            skip_pyc_compilation: ConditionalList(
                                [],
                            ),
                            use_python_app_entrypoint: false,
                            version_independent: false,
                            site_packages_path: None,
                        },
                        skip: ConditionalList(
                            [],
                        ),
                        always_copy_files: ConditionalList(
                            [],
                        ),
                        always_include_files: ConditionalList(
                            [],
                        ),
                        merge_build_and_host_envs: false,
                        files: List(
                            ConditionalList(
                                [],
                            ),
                        ),
                        dynamic_linking: DynamicLinking {
                            rpaths: ConditionalList(
                                [],
                            ),
                            binary_relocation: Boolean(
                                Concrete {
                                    value: true,
                                    span: Span {
                                        start: 0,
                                        end: 0,
                                        start_line: 0,
                                        start_column: 0,
                                        end_line: 0,
                                        end_column: 0,
                                    },
                                },
                            ),
                            missing_dso_allowlist: ConditionalList(
                                [],
                            ),
                            rpath_allowlist: ConditionalList(
                                [],
                            ),
                            overdepending_behavior: None,
                            overlinking_behavior: None,
                        },
                        variant: VariantKeyUsage {
                            use_keys: ConditionalList(
                                [],
                            ),
                            ignore_keys: ConditionalList(
                                [],
                            ),
                            down_prioritize_variant: None,
                        },
                        prefix_detection: PrefixDetection {
                            force_file_type: ForceFileType {
                                text: ConditionalList(
                                    [],
                                ),
                                binary: ConditionalList(
                                    [],
                                ),
                            },
                            ignore: Boolean(
                                Concrete {
                                    value: false,
                                    span: Span {
                                        start: 0,
                                        end: 0,
                                        start_line: 0,
                                        start_column: 0,
                                        end_line: 0,
                                        end_column: 0,
                                    },
                                },
                            ),
                            ignore_binary_files: false,
                        },
                        post_process: [],
                    },
                    about: About {
                        homepage: None,
                        license: None,
                        license_file: ConditionalList(
                            [],
                        ),
                        license_family: None,
                        summary: Some(
                            Concrete {
                                value: "Sub package also inheriting from top-level",
                                span: Span {
                                    start: 955,
                                    end: 955,
                                    start_line: 45,
                                    start_column: 16,
                                    end_line: 45,
                                    end_column: 16,
                                },
                            },
                        ),
                        description: None,
                        documentation: None,
                        repository: None,
                    },
                    tests: [
                        Commands(
                            CommandsTest {
                                script: ConditionalList(
                                    [
                                        Value(Concrete { value: Command("python -c \"import sub_package\""), span: Span { start: 1037, end: 1037, start_line: 48, start_column: 13, end_line: 48, end_column: 13 } }),
                                    ],
                                ),
                                requirements: None,
                                files: None,
                            },
                        ),
                    ],
                },
            ),
            Package(
                PackageOutput {
                    package: PackageMetadata {
                        name: Concrete {
                            value: PackageName(
                                PackageName {
                                    normalized: None,
                                    source: "custom-package",
                                },
                            ),
                            span: Span {
                                start: 1155,
                                end: 1155,
                                start_line: 52,
                                start_column: 13,
                                end_line: 52,
                                end_column: 13,
                            },
                        },
                        version: Some(
                            Concrete {
                                value: VersionWithSource {
                                    version: Version {
                                        version: [[0], [5], [0], [0]],
                                        local: [],
                                    },
                                    source: Some(
                                        "5.0.0",
                                    ),
                                },
                                span: Span {
                                    start: 1185,
                                    end: 1185,
                                    start_line: 53,
                                    start_column: 16,
                                    end_line: 53,
                                    end_column: 16,
                                },
                            },
                        ),
                    },
                    inherit: TopLevel,
                    source: [],
                    requirements: Requirements {
                        build: ConditionalList(
                            [],
                        ),
                        host: ConditionalList(
                            [],
                        ),
                        run: ConditionalList(
                            [
                                Value(Concrete { value: SerializableMatchSpec(MatchSpec { name: Some(PackageName { normalized: None, source: "python" }), version: None, build: None, build_number: None, file_name: None, extras: None, channel: None, subdir: None, namespace: None, md5: None, sha256: None, url: None, license: None }), span: Span { start: 1422, end: 1422, start_line: 64, start_column: 11, end_line: 64, end_column: 11 } }),
                            ],
                        ),
                        run_constraints: ConditionalList(
                            [],
                        ),
                        run_exports: RunExports {
                            noarch: ConditionalList(
                                [],
                            ),
                            strong: ConditionalList(
                                [],
                            ),
                            strong_constraints: ConditionalList(
                                [],
                            ),
                            weak: ConditionalList(
                                [],
                            ),
                            weak_constraints: ConditionalList(
                                [],
                            ),
                        },
                        ignore_run_exports: IgnoreRunExports {
                            by_name: ConditionalList(
                                [],
                            ),
                            from_package: ConditionalList(
                                [],
                            ),
                        },
                    },
                    build: Build {
                        number: Concrete {
                            value: 1,
                            span: Span {
                                start: 1234,
                                end: 1234,
                                start_line: 56,
                                start_column: 15,
                                end_line: 56,
                                end_column: 15,
                            },
                        },
                        string: None,
                        script: ConditionalList(
                            [
                                Value(Concrete { value: Command("echo \"Custom build overriding top-level\""), span: Span { start: 1260, end: 1260, start_line: 58, start_column: 11, end_line: 58, end_column: 11 } }),
                            ],
                        ),
                        noarch: None,
                        python: PythonBuild {
                            entry_points: ConditionalList(
                                [],
                            ),
                            skip_pyc_compilation: ConditionalList(
                                [],
                            ),
                            use_python_app_entrypoint: false,
                            version_independent: false,
                            site_packages_path: None,
                        },
                        skip: ConditionalList(
                            [],
                        ),
                        always_copy_files: ConditionalList(
                            [],
                        ),
                        always_include_files: ConditionalList(
                            [],
                        ),
                        merge_build_and_host_envs: false,
                        files: List(
                            ConditionalList(
                                [],
                            ),
                        ),
                        dynamic_linking: DynamicLinking {
                            rpaths: ConditionalList(
                                [],
                            ),
                            binary_relocation: Boolean(
                                Concrete {
                                    value: true,
                                    span: Span {
                                        start: 0,
                                        end: 0,
                                        start_line: 0,
                                        start_column: 0,
                                        end_line: 0,
                                        end_column: 0,
                                    },
                                },
                            ),
                            missing_dso_allowlist: ConditionalList(
                                [],
                            ),
                            rpath_allowlist: ConditionalList(
                                [],
                            ),
                            overdepending_behavior: None,
                            overlinking_behavior: None,
                        },
                        variant: VariantKeyUsage {
                            use_keys: ConditionalList(
                                [],
                            ),
                            ignore_keys: ConditionalList(
                                [],
                            ),
                            down_prioritize_variant: None,
                        },
                        prefix_detection: PrefixDetection {
                            force_file_type: ForceFileType {
                                text: ConditionalList(
                                    [],
                                ),
                                binary: ConditionalList(
                                    [],
                                ),
                            },
                            ignore: Boolean(
                                Concrete {
                                    value: false,
                                    span: Span {
                                        start: 0,
                                        end: 0,
                                        start_line: 0,
                                        start_column: 0,
                                        end_line: 0,
                                        end_column: 0,
                                    },
                                },
                            ),
                            ignore_binary_files: false,
                        },
                        post_process: [],
                    },
                    about: About {
                        homepage: None,
                        license: Some(
                            Concrete {
                                value: License(
                                    MIT,
                                ),
                                span: Span {
                                    start: 1379,
                                    end: 1379,
                                    start_line: 61,
                                    start_column: 16,
                                    end_line: 61,
                                    end_column: 16,
                                },
                            },
                        ),
                        license_file: ConditionalList(
                            [],
                        ),
                        license_family: None,
                        summary: Some(
                            Concrete {
                                value: "Custom package with overridden build",
                                span: Span {
                                    start: 1327,
                                    end: 1327,
                                    start_line: 60,
                                    start_column: 16,
                                    end_line: 60,
                                    end_column: 16,
                                },
                            },
                        ),
                        description: None,
                        documentation: None,
                        repository: None,
                    },
                    tests: [],
                },
            ),
        ],
    },
)
